### DISCUSSION TITLE
EGSnrc MAC OS - library not loaded: /usr/lib/libSystem.B.dylib

### DISCUSSION THREAD

**Discussion by hbouchard** [nrc-cnrc/EGSnrc#963] [Questions] [answered]
Hello EGSperts, 

Some in my class are running EGSnrc on MAC OS and we are getting library not loaded: `/usr/lib/libSystem.B.dylib` error.

We experience this issue on MacBook Pros M1 8GB RAM (Monterey and Ventura) when changing `$MAXRADII` from 4 to 16 in flurznrc. It compiles fine, but when attempting to run I get the following error message:

```
dyld[18643]: dyld cache '(null)' not loaded: syscall to map cache into shared region failed
dyld[18643]: Library not loaded: /usr/lib/libSystem.B.dylib
  Referenced from: <C4F425D6-E281-31CB-9A2B-2FB3103A1A67> /Users/hbouchard/EGSnrc/egs_home/bin/osx/flurznrc
  Reason: tried: '/Users/hbouchard/EGSnrc/HEN_HOUSE/egs++/dso/osx/libSystem.B.dylib' (no such file), '/Users/hbouchard/EGSnrc/HEN_HOUSE/egs++/dso/osx/libSystem.B.dylib' (no such file), '/libSystem.B.dylib' (no such file), '/usr/lib/libSystem.B.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/lib/libSystem.B.dylib' (no such file), '/usr/lib/libSystem.B.dylib' (no such file, no dyld cache), '/usr/local/lib/libSystem.B.dylib' (no such file)
Abort trap: 6 
```

It seems quite serious? I have read people complaining about this issue without apparent solution. This problem doesn't occur with my iMac Intel i7 16GB RAM.

Many thanks!

**Comment by hbouchard** [comment]
Just to add to the discussion, here a recent post I found with no answer:

https://developer.apple.com/forums/thread/722360

**Comment by ftessier** [comment]
@mchamberland (the expert with EGSnrc on M1!), does that ring a bell?

**Comment by ftessier** [comment, marked as answer]
@hbouchard, so it runs fine when `$MAXRADII` is 4?

**Reply by rtownson** [reply]
@hbouchard If this is the case, try reducing $MAXZREG (if that's acceptable) to offset the increased memory usage. Decreasing $MXSTACK might also help a touch.

**Reply by ftessier** [reply]
Good point, is this a memory issue? What is the total number of slabs, i.e., the total number of regions here?

**Reply by hbouchard** [reply]
Yes fine with `$MAXRADII=4`

**Reply by rtownson** [reply]
Just an idea, but at the end of the flurznrc `Makefile`, try adding this line, and checking `$MAXRADII` as 16 again:

```
FOPT += -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE
```

**Reply by mchamberland** [reply]
@rtownson 
> Just an idea, but at the end of the flurznrc `Makefile`, try adding this line, and checking `$MAXRADII` as 16 again:
> 
> ```
> FOPT += -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE
> ```

This did not resolve the issue for me.

**Reply by rtownson** [reply]
@mchamberland how about `-mcmodel=large`?

**Reply by mchamberland** [reply]
@rtownson  Nope. That flag makes compilation fail with a bunch of errors like so:

```
/var/folders/0j/fkb5fbkn3m3_xdmlbxl7371m0000gn/T//ccZGOudX.s:169285:18: error: invalid variant 'BLEAH'
        adrp    x0, lC6789@BLEAH
```

**Reply by ftessier** [reply]
Great, @rtownson found the answer, but I get credit! 😄

**Comment by ftessier** [comment]
What is the output of `echo $DYLD_LIBRARY_PATH` in your Terminal?

**Reply by ftessier** [reply]
It looks as though your dynamic lilbrary path is trying to load a system library from the egs++ directory (?). You could try `DYLD_LIBRARY_PATH=` in your shell, you don't need egs++ for flurznrc.

**Reply by hbouchard** [reply]
echo $DYLD_LIBRARY_PATH
gives
/Users/hbouchard/EGSnrc/HEN_HOUSE/egs++/dso/osx:/Users/hbouchard/EGSnrc/HEN_HOUSE/egs++/dso/osx:

**Reply by hbouchard** [reply]
> It looks as though your dynamic lilbrary path is trying to load a system library from the egs++ directory (?). You could try `DYLD_LIBRARY_PATH=` in your shell, you don't need egs++ for flurznrc.

When I run after that command, I get this:

dyld[18911]: dyld cache '(null)' not loaded: syscall to map cache into shared region failed
dyld[18911]: Library not loaded: ../egs++/dso/osx/libiaea_phsp.dylib
  Referenced from: <C4F425D6-E281-31CB-9A2B-2FB3103A1A67> /Users/hbouchard/EGSnrc/egs_home/bin/osx/flurznrc
  Reason: tried: '/libiaea_phsp.dylib' (no such file), '../egs++/dso/osx/libiaea_phsp.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS../egs++/dso/osx/libiaea_phsp.dylib' (no such file), '../egs++/dso/osx/libiaea_phsp.dylib' (no such file), '/usr/local/lib/libiaea_phsp.dylib' (no such file), '/usr/lib/libiaea_phsp.dylib' (no such file, no dyld cache)
Abort trap: 6

**Reply by ftessier** [reply]
Évidemment le linker ne trouve plus les biliothèques de EGSnrc, dont celle pour les fichiers iaea 🤦🏻‍♂️. Peux-tu essayer d'ajouter l'emplacement de la bibliotèque `libSsytem.B.dylib` au `$DYLD_LIBRARY_PATH`. Cherche ce nom de fichier pour obtenir un emplacement comme `/application/Xcode/(...)/libSsytem.B.dylib` (tu peux aussi tenter `locate libSsytem.B.dylib` dans le Terminal). Ensuite ajoute cet emplacement dans la variable d'environnement:

```bash
DYLD_LIBRARY_PATH=$DYLD_LIBRARY_PATH:/application/Xcode/(...)/libSsytem.B.dylib
```

Référence: https://developer.apple.com/forums/thread/669533 (mais ça date; je crois que `libSsytem.B.dylib` n'existe plus en tant que fichier sur le disque).

**Reply by mchamberland** [reply]
@ftessier  Indeed, libSystem.B.dylib is no longer shipped on Apple Silicon. Well... there is _a_ libSystem.B.dylib, but it is for iOS, not macOS:

`'/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib//libSystem.B.dylib' (mach-o file (/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib//libSystem.B.dylib), but incompatible platform (have 'iOS-sim', need 'macOS'))`

**Comment by mchamberland** [comment]
@ftessier I have not encountered this before, but I will try reproducing.

**Comment by ftessier** [comment]
You might want to double-check if you have any `-mcmodel` flag (see #905), although this usually causes compilation failure at the linking stage.

**Comment by mchamberland** [comment]
@ftessier Indeed, I can reproduce on an M2 Mac with macOS Ventura 13.2. Changing $MAXRADII to 16 causes the error upon running the application. I have 24 GB of RAM. I don't see any `-mcmodel` flag in my logs.

**Comment by mchamberland** [comment]
@ftessier @rtownson Reducing $MAXZREG indeed seems to resolve the issue.

**Comment by ftessier** [comment]
Thank you @mchamberland, I am promoting this discussion to an issue.

**Comment by ftessier** [comment]
Is this specific to flurznrc? I presume not.

**Reply by hbouchard** [reply]
The default for dosrznrc is `$MAXZREG=200` and `$MAXRADII=60`. With the same settings I had initially with flurznrc, i.e., `$MAXZREG=500` and `$MAXRADII=16`, the dosrznrc simulation works fine. I guess the number of scoring arrays are different.

**Comment by hbouchard** [comment]
Since `$MXREG={COMPUTE $MAXRADII*($MAXZPLANE-1) +1}`, I tried reducing `$MAXZREG` from 500 to 200. It worked fine.

**Comment by ftessier** [comment]
See #964 for the explanation.
